connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java

```diff
@@ -183,12 +187,18 @@ public MirrorClientConfig clientConfig(String cluster) {
 
         // fill in reasonable defaults
         props.putIfAbsent(GROUP_ID_CONFIG, sourceAndTarget.source() + "-mm2");
-        props.putIfAbsent(DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG, "mm2-offsets."
-                + sourceAndTarget.source() + ".internal");
-        props.putIfAbsent(DistributedConfig.STATUS_STORAGE_TOPIC_CONFIG, "mm2-status."
-                + sourceAndTarget.source() + ".internal");
-        props.putIfAbsent(DistributedConfig.CONFIG_TOPIC_CONFIG, "mm2-configs."
-                + sourceAndTarget.source() + ".internal");
+
+        String separator = originalsStrings().getOrDefault(REPLICATION_POLICY_SEPARATOR, REPLICATION_POLICY_SEPARATOR_DEFAULT);
+        if (separator.equals("-")) {
+            throw new ConfigException("You should not use a single dash as a " + REPLICATION_POLICY_SEPARATOR);
+        }
+
+        props.putIfAbsent(DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG, "mm2-offsets" + separator
```

Comment:
> Oh no, this lines replace the original props.putIfAbsent(DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG, "mm2-offsets." + sourceAndTarget.source() + ".internal");, etc. not Connect's internal topics.

`DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG` is one of the connect's internal topics. 
```
private static final String OFFSET_STORAGE_TOPIC_CONFIG_DOC = "The name of the Kafka topic where connector offsets are stored";
```

My point is customers already can control these types of topics using the `DistributedConfig` so there's no point in controlling them again using the seprator. The main issue I think we need to fix first is preventing is the replication of these topics.

Reply:
> DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG is one of the connect's internal topics.

Oh, sorry for my misunderstanding; Yes, you are right; `DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG` is one of Kafka Connect's internal topics but, it is created as `mm2-offsets.{source}.internal` **only when MM2 is running in standalone mode.** (see: `MirrorMakerConfig.workerConfig`) My apologies.

If the user is running MM2 in connect mode, the user is responsible for configuring `DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG`, `DistributedConfig.CONFIG_TOPIC_CONFIG`, etc. It is what you are meaning. Right?

Reply:
Yes, most production use-cases will be using MM2 with connect mode. We just need to protect users who either didn't bother by setting up their connect topics, or those who run standalone mode.

Reply:
Yes. The main problem am seeing is preventing the unintentional replication of these topics for users who either didn't set up their connect's topics or those who run standalone/dedicated cluster mode.

Reply:
> They are created in any mode if there is no value for `DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG`

Yes, I agree. I mean, it is created as `mm2-offsets.{source}.internal` if run in standalone mode, and `DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG` is not explicitly set. (You know, there is a default value of it in Kafka Connect.)

> The PR's approach is trying to control the Connect topics that MM2 needs to set up using the separator.

Totally agree.

> This is where I am not sure it's a minor fix or something that requires a KIP that follows KIP-690.

Yes, I am also not sure whether this issue is fixing an unintended, omitted function or proposing a new interface. @mimaison @ryannedolan How do you think?

Reply:
My bad, I missed the fact that this line is about the Kafka Connect's internal topics and not MM2's. I think it's fine to keep the code as is.

Reply:
I have also thought over this issue again.

The core of the problem seems to be, there is some gap in our configuration policy, and it causes the Kafka Connect internal topics may be mirrored into the other cluster. At first, I thought it could only happen in standalone mode, but it is not true - my apologies for the confusion.

There are two cases this problem may happen:

1. Using a custom separator in standalone mode after KIP-690 (i.e., AK 3.1.0), what I and @OmniaGM discussed until now.
2. Running MM2 in Kafka Connect with the default `[config, offset, status].storage.topic` configuration. (what @mimaison just pointed out)

2 may be prevented by `topics` configuration override, and it seems like the users are already doing so. (Yes, it has been like this since day 1.) So, let's focus on 1 only.

To prevent those topics from being mirrored with a custom separator, there are three approaches:

a. Fix `ReplicationPolicy#isInternalTopic` only.
b. Leave the users to override `topics` configuration like Kafka Connect mode.
c. Make `mm2-[offsets, status, configs].{source}.internal` topics to use given separator.

Approach a seems unavailable; to determine whether given `mm2-[offsets, status, configs].{source}.internal` topic is an internal one, `ReplicationPolicy#isInternalTopic` method should know the source cluster's alias, since those topics include `{source}` in their name. But actually, it takes a topic name as a parameter only.

Approach b is good, but it may be not very clear for the users with little experience.

Removing approach a and b, only c survives - and it is why I thought applying the separator to those topic names is the only way to make it work identically with the default separator case (which does not mirror those topics by default).

How do you think?

Reply:
One issue with c. is that it works for new environments. If users already have MM2 running, it's using topics with the current names.

Reply:
> One issue with c. is that it works for new environments. If users already have MM2 running, it's using topics with the current names.

Yes, I already thought of this backward compatibility issue - how about this?: **"If the user is using a custom separator, no MM2 internal topics are created yet, and there are already default MM2 internal topics, migrate the default MM2 internal topic contents into the new one."**

With this strategy, the users already running MM2 can change the configuration without any issues and manual work.

Plus, it is also good for the `MirrorClient` to export/import the MM2 internal topics. This strategy is not automatic but, it gives the users more flexibility - it handles the separator change from a custom one to the other custom one.

How about these approaches?

Reply:
I think the fix @OmniaGM proposed above (https://github.com/apache/kafka/pull/11431#discussion_r737592730) makes sense and I think it's fully compatible. WDYT?

Reply:
Okay. Thanks for your patience in convincing me.

Then, 1) Making the default internal topic names (e.g., `DistributedConfig.OFFSET_STORAGE_TOPIC_CONFIG`) in the standalone mode consistent 2) Provide a way to migrate internal topics are beyond the scope of this issue, and it should be filed as a separate KIP. Right?